app/src/main/java/com/github/featuredetectandroid/MainActivity.kt

```diff
@@ -60,62 +48,57 @@ class MainActivity : ComponentActivity() {
+            height = outputViewModel.featureDetector?.height ?: 0
+        )
 
         setContent {
             FeatureDetectAppTheme {
                 var selectedAlgorithm by remember {
                     mutableStateOf(preferencesManager.getSelectedAlgorithm())
                 }
-                Scaffold(
-                    drawerContent = {
-                        Menu(
-                            header = "Keypoint detection algorithm:",
-                            options = KeypointDetectionAlgorithm.names,
-                            selectedOption = selectedAlgorithm,
-                            onSelected = { algorithmName ->
-                                preferencesManager.putSelectedAlgorithm(algorithmName)
-                                selectedAlgorithm = algorithmName
-                            }
-                        )
-                    }
-                ) {
-                    Box(Modifier.fillMaxSize()) {
-                        if (imageViewModel.isCameraPermissionGranted) {
-                            imageViewModel.grayscaleBitmap?.let { bitmap ->
-                                Image(
-                                    bitmap = bitmap.asImageBitmap(),
-                                    modifier = Modifier.fillMaxSize(),
-                                    contentDescription = "Grayscale photo"
-                                )
-                            }
-                        } else {
-                            Text(
-                                text = "Camera permission required",
-                                modifier = Modifier.align(Alignment.Center)
+                AppLayout(
+                    isCameraPermissionGranted = isCameraPermissionGranted(),
+                    keypointOffsets = outputViewModel.keypointOffsets,
+                    frameBitmap = outputViewModel.frameBitmap,
+                    selectedAlgorithm = selectedAlgorithm,
+                    onAlgorithmSelected = { algorithmName ->
+                        preferencesManager.putSelectedAlgorithm(algorithmName)
+                        selectedAlgorithm = algorithmName
+                        outputViewModel.keypointOffsets = emptyList()
+                        outputViewModel.featureDetector =
+                            KeypointDetectionAlgorithm.nameToFeatureDetector(
+                                algorithmName = preferencesManager.getSelectedAlgorithm(),
+                                context = this,
+                                width = outputViewModel.featureDetector?.width ?: 0,
+                                height = outputViewModel.featureDetector?.height ?: 0
                             )
-                        }
                     }
-                }
+                )
             }
         }
     }
 
     override fun onResume() {
         super.onResume()
-        imageViewModel.isCameraPermissionGranted = isCameraPermissionGranted()
         tryStartCamera()
```

Comment:
We always start the camera but never stop it in `onPause` (actually, I would put it in in `onStart` and `onStop` respectively). I see that it calls `unbindAll` before starting, so nothing bad happen, but let's create an issue about this and add later -- no need to listen to the camera when the activity isn't visible.

Reply:
I created the issue for it.